### PR TITLE
WIP: Integrate FIP0013

### PR DESCRIPTION
**Pull request by arajasek**
Overhauls #5769 to be on v5 actors and go into the staging branch.

Currently this is pulling in https://github.com/filecoin-project/filecoin-ffi/pull/169, https://github.com/filecoin-project/go-paramfetch/pull/13 and https://github.com/filecoin-project/specs-actors/pull/1381, and implementing some plumbing logic for proof aggregation

TODO:
- [ ] Write some unit tests for CommitBatcher
- [ ] Write an integration test for commit batching
  - [ ] Test edge-cases
    - [ ] One porep fails
    - [ ] One commit expires? (deal start?)
    - [ ] Batch fails for whatever reason (e.g. out of gas)
- [ ] Use correct specs-actors commit (currently using `asr/temp`).

### CODE REVIEW

**Comment by BigLep** [comment]
2021-05-13 conversation:
1. Around FFIs: https://github.com/filecoin-project/go-state-types/commit/4aaed5f37509899cb438a761b501808ac0212529#diff-1db57c219b5669cc19920de5e805ebd15be6809e27771250bced48416867d80dR103
2. Where proofs are coming from:
 - Are regular PoReps 
 - We don't think we need a new API.  We "think" the logic is in the miner state machine.

**Comment by BigLep** [comment]
2021-05-17: for getting this to merge:
1. configuration
2. unit tests

File: node/config/def.go [hunk 1]
```diff
@@ -237,6 +241,10 @@ func DefaultStorageMiner() *StorageMiner {
 			MaxSealingSectorsForDeals: 0,
 			WaitDealsDelay:            Duration(time.Hour * 6),
 			AlwaysKeepUnsealedCopy:    true,
+
+			AggregateCommits: true,
+			MinCommitBatch:   5, // todo: base this on some real numbers
+			MaxCommitBatch:   400,
```
Comment (vyzo) [inline node/config/def.go:247]: this cannot exceed 204 per FIP.

File: extern/sector-storage/ffiwrapper/types.go [hunk 2]
```diff
@@ -34,11 +34,15 @@ type Storage interface {
...
-	VerifyWinningPoSt(ctx context.Context, info proof2.WinningPoStVerifyInfo) (bool, error)
-	VerifyWindowPoSt(ctx context.Context, info proof2.WindowPoStVerifyInfo) (bool, error)
+	VerifySeal(proof5.SealVerifyInfo) (bool, error)
+	VerifyAggregateSeals(aggregate proof5.AggregateSealVerifyProofAndInfos) (bool, error)
+	VerifyWinningPoSt(ctx context.Context, info proof5.WinningPoStVerifyInfo) (bool, error)
+	VerifyWindowPoSt(ctx context.Context, info proof5.WindowPoStVerifyInfo) (bool, error)
 
 	GenerateWinningPoStSectorChallenge(context.Context, abi.RegisteredPoStProof, abi.ActorID, abi.PoStRandomness, uint64) ([]uint64, error)
+
+	// cheap, makes no sense to put this on the storage interface
```
Comment (Stebalien) [inline extern/sector-storage/ffiwrapper/types.go:44]: Really, neither this nor `GenerateWinningPoStSectorChallenge` belong here. Why do this?

File: extern/storage-sealing/commit_batch.go [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+	}
+
+	if after && total < cfg.MinCommitBatch {
+		return nil, nil
+	}
+
+	spt := b.todo[0].spt
+	proofs := make([][]byte, 0, total)
+
+	for id, p := range b.todo {
```
Comment (GFZRZK) [inline extern/storage-sealing/commit_batch.go:187]: here can take at most `MaxAggregatedSectors` and proof size can't longer than `MaxAggregateProofSize`, if you don't defend here the `processBatch` may never work out a right result.

File: extern/sector-storage/ffiwrapper/types.go (hunk 2 above)
Comment (magik6k) [reply extern/sector-storage/ffiwrapper/types.go:44]: Yeah, we probably want a separate interface for these. This is here because the Sealer instance is wrapped into a Worker+SealingMgr, and it really makes no sense to put those 2 calls behind those layers of abstraction.

The `Verifier` interface is the other ready-to-use interface that's already plumbed into all the places we need those calls in.

File: extern/storage-sealing/commit_batch.go [hunk 4]
```diff
@@ -0,0 +1,363 @@
...
+		params.SectorNumbers.Set(uint64(id))
+		infos = append(infos, p.info)
+	}
+
+	sort.Slice(infos, func(i, j int) bool {
+		return infos[i].Number < infos[j].Number
+	})
+
+	for _, info := range infos {
+		if len(infos) >= cfg.MaxCommitBatch {
```
Comment (DC-Byteark) [inline extern/storage-sealing/commit_batch.go:197]: This should be proofs, not infos.

File: api/test/window_post.go [hunk 5]
```diff
@@ -117,6 +117,52 @@ func TestSDRUpgrade(t *testing.T, b APIBuilder, blocktime time.Duration) {
...
+
+	for {
+		h, err := client.ChainHead(ctx)
+		require.NoError(t, err)
+		if h.Height() > 10 {
+			break
+		}
+	}
+
+	pledgeSectors(t, ctx, miner, nSectors, 0, nil)
```
Comment (magik6k) [inline api/test/window_post.go:160]: todo more asserts here

File: api/test/window_post.go [hunk 6]
```diff
@@ -117,6 +117,52 @@ func TestSDRUpgrade(t *testing.T, b APIBuilder, blocktime time.Duration) {
 	<-done
 }
 
+func TestPledgeBatching(t *testing.T, b APIBuilder, blocktime time.Duration, nSectors int) {
```
Comment (magik6k) [inline api/test/window_post.go:120]: Should be in a different file

File: extern/storage-sealing/states_sealing.go [hunk 7]
```diff
@@ -224,56 +225,50 @@ func (m *Sealing) remarkForUpgrade(sid abi.SectorNumber) {
 	}
 }
 
-func (m *Sealing) handlePreCommitting(ctx statemachine.Context, sector SectorInfo) error {
+func (m *Sealing) preCommitParams(ctx statemachine.Context, sector SectorInfo) (*miner.SectorPreCommitInfo, big.Int, TipSetToken, error) {
```
Comment (magik6k) [inline extern/storage-sealing/states_sealing.go:228]: This can send events, but we ignore that in callers

File: storage/wdpost_run_test.go [hunk 8]
```diff
@@ -144,6 +145,14 @@ func (m mockVerif) VerifyWindowPoSt(ctx context.Context, info proof2.WindowPoStV
 	return true, nil
 }
 
+func (m mockVerif) AggregateSealProofs(proofType abi.RegisteredSealProof, rap abi.RegisteredAggregationProof, proofs [][]byte) ([]byte, error) {
+	panic("implement me")
+}
```
Comment (magik6k) [inline storage/wdpost_run_test.go:150]: Not on the interface anymore

File: documentation/en/api-methods-miner.md [hunk 9]
```diff
@@ -1556,6 +1560,24 @@ Response: `{}`
 ## Sector
 
 
+### SectorCommitFlush
```
Comment (ZenGround0) [inline documentation/en/api-methods-miner.md:1563]: nit: add docs

File: documentation/en/cli-lotus-miner.md [hunk 10]
```diff
@@ -0,0 +1,1864 @@
+# lotus-miner
```
Comment (ZenGround0) [inline documentation/en/cli-lotus-miner.md:1]: Most landed in #6145, only reviewing batching section which lgtm

File: extern/sector-storage/ffiwrapper/types.go [hunk 11]
```diff
@@ -34,13 +34,21 @@ type Storage interface {
...
+	VerifyAggregateSeals(aggregate proof5.AggregateSealVerifyProofAndInfos) (bool, error)
+	VerifyWinningPoSt(ctx context.Context, info proof5.WinningPoStVerifyInfo) (bool, error)
+	VerifyWindowPoSt(ctx context.Context, info proof5.WindowPoStVerifyInfo) (bool, error)
 
 	GenerateWinningPoStSectorChallenge(context.Context, abi.RegisteredPoStProof, abi.ActorID, abi.PoStRandomness, uint64) ([]uint64, error)
 }
 
+// Prover contains cheap proving-related methods
+type Prover interface {
+	// TODO: move GenerateWinningPoStSectorChallenge from the Verifier interface to here
```
Comment (ZenGround0) [inline extern/sector-storage/ffiwrapper/types.go:47]: Reminder to do this

File: extern/sector-storage/mock/mock.go [hunk 12]
```diff
@@ -509,12 +511,49 @@ func (m mockVerif) VerifySeal(svi proof2.SealVerifyInfo) (bool, error) {
...
+			b *= uint8(pi) // with aggregate index
+			out[i] += b
+		}
+	}
+
+	var sis []abi.SectorNumber
+	for _, info := range aggregate.Infos {
+		sis = append(sis, info.Number)
+	}
+	fmt.Printf("VERSIS %+v\n", sis)
```
Comment (ZenGround0) [inline extern/sector-storage/mock/mock.go:529]: nit: possible debug print

File: extern/sector-storage/mock/mock.go [hunk 13]
```diff
@@ -509,12 +511,49 @@ func (m mockVerif) VerifySeal(svi proof2.SealVerifyInfo) (bool, error) {
...
+		for i := range proof[:32] {
+			out[i] += proof[i] * uint8(pi)
+		}
+	}
+
+	var sis []abi.SectorNumber
+	for _, info := range aggregateInfo.Infos {
+		sis = append(sis, info.Number)
+	}
+	fmt.Printf("AGGSIS %+v\n", sis)
```
Comment (ZenGround0) [inline extern/sector-storage/mock/mock.go:546]: nit: possible debug print

File: markets/storageadapter/ondealsectorcommitted.go [hunk 14]
```diff
@@ -137,26 +138,47 @@ func (mgr *SectorCommittedManager) OnDealSectorPreCommitted(ctx context.Context,
...
-		// Check through the deal IDs associated with this message
-		for _, did := range params.DealIDs {
-			if did == res.DealID {
-				// Found the deal ID in this message. Callback with the sector ID.
-				cb(params.SectorNumber, false, nil)
-				return false, nil
+		// Extract the message parameters
+		switch msg.Method {
+		case miner.Methods.PreCommitSector:
+			var params miner.SectorPreCommitInfo
```
Comment (ZenGround0) [inline markets/storageadapter/ondealsectorcommitted.go:151]: suggestion: this feels like it belongs in a function `onSectorNumber(msg.Methods, msg.Params, res.DealID, cb)`

File: markets/storageadapter/ondealsectorcommitted.go [hunk 15]
```diff
@@ -207,16 +229,35 @@ func (mgr *SectorCommittedManager) OnDealSectorCommitted(ctx context.Context, pr
...
-		if msg.To != provider || msg.Method != miner.Methods.ProveCommitSector {
+		if msg.To != provider {
 			return false, nil
 		}
 
-		var params miner.ProveCommitSectorParams
-		if err := params.UnmarshalCBOR(bytes.NewReader(msg.Params)); err != nil {
-			return false, xerrors.Errorf("failed to unmarshal prove commit sector params: %w", err)
-		}
+		switch msg.Method {
```
Comment (ZenGround0) [inline markets/storageadapter/ondealsectorcommitted.go:236]: again consider putting this in its own function to keep `OnDealSectorCommitted` more compact

File: node/config/def.go [hunk 16]
```diff
@@ -82,6 +84,22 @@ type SealingConfig struct {
 
 	AlwaysKeepUnsealedCopy bool
 
+	BatchPreCommits     bool
+	MaxPreCommitBatch   int
+	MinPreCommitBatch   int
+	PreCommitBatchWait  Duration
+	PreCommitBatchSlack Duration
```
Comment (ZenGround0) [inline node/config/def.go:91]: Are there docs describing these things?  I have a general idea what each field does from the names but it's not exactly obvious.

**Review by ZenGround0 — COMMENTED** [review]
About half way

File: extern/storage-sealing/commit_batch.go [hunk 17]
```diff
@@ -0,0 +1,363 @@
...
+
+	for {
+		if forceRes != nil {
+			forceRes <- lastMsg
+			forceRes = nil
+		}
+		lastMsg = nil
+
+		var sendAboveMax, sendAboveMin bool
+		select {
```
Comment (vyzo) [inline extern/storage-sealing/commit_batch.go:102]: I think we should only start the timer after the first commit enters the queue.

File: extern/sector-storage/ffiwrapper/types.go (hunk 11 above)
Comment (magik6k) [reply extern/sector-storage/ffiwrapper/types.go:47]: Should be done outside of this PR after it lands

Comment (magik6k) [reply extern/sector-storage/ffiwrapper/types.go:47]: (warrants an issue)

File: chain/stmgr/forks.go [hunk 18]
```diff
@@ -1247,7 +1247,7 @@ func upgradeActorsV5Common(
 
 	// Persist the result.
 	newRoot, err := store.Put(ctx, &types.StateRoot{
-		Version: types.StateTreeVersion3,
+		Version: types.StateTreeVersion4,
```
[the remark this answers is not in the document]
Comment (magik6k) [reply chain/stmgr/forks.go:1250]: Oh, I just assumed that /something/ is expecting this to be bumped based on the nv12 diff, might have been wrong

File: extern/sector-storage/ffiwrapper/types.go (hunk 11 above)
Comment (magik6k) [reply extern/sector-storage/ffiwrapper/types.go:47]: https://github.com/filecoin-project/lotus/issues/6326

File: chain/vm/gas.go [hunk 19]
```diff
@@ -158,7 +160,8 @@ var prices = map[abi.ChainEpoch]Pricelist{
 
 		hashingBase:                  31355,
 		computeUnsealedSectorCidBase: 98647,
-		verifySealBase:               2000, // TODO gas , it VerifySeal syscall is not used
+		verifySealBase:               2000,        // TODO gas , it VerifySeal syscall is not used
```
[the remark this answers is not in the document]
Comment (magik6k) [reply chain/vm/gas.go:163]: Unless @Kubuxu disagrees I'd say no

File: chain/stmgr/forks.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (magik6k) [reply chain/stmgr/forks.go:1250]: I guess the only good reason for it is version check in `chain/stmgr/forks.go:upgradeActorsV5Common`, I'm ok with dropping that

**Review by ZenGround0 — APPROVED** [review]
Integration LGTM

I have not yet reviewed the extensive storage-sealing changes.  If we want a thorough review from me on this it is going to take significant time to get up to speed because the storage fsm is totally new code to me.  That might be expected but I want to check in about it before moving forward.

File: extern/storage-sealing/commit_batch.go [hunk 20]
```diff
@@ -0,0 +1,367 @@
...
+
+type CommitBatcherApi interface {
+	SendMsg(ctx context.Context, from, to address.Address, method abi.MethodNum, value, maxFee abi.TokenAmount, params []byte) (cid.Cid, error)
+	StateMinerInfo(context.Context, address.Address, TipSetToken) (miner.MinerInfo, error)
+	ChainHead(ctx context.Context) (TipSetToken, abi.ChainEpoch, error)
+}
+
+type AggregateInput struct {
+	spt abi.RegisteredSealProof
+	// TODO: Something changed in actors, I think this now needs to be AggregateSealVerifyProofAndInfos todo ??
```
Comment (arajasek) [inline extern/storage-sealing/commit_batch.go:37]: This can now be removed

**Review by arajasek — COMMENTED** [review]
Reviewed the storage logic, and it looks good. Ran a devnet and it seems to do the thing.

(Verbal green check mark from me, since it's """my PR""")

File: chain/stmgr/forks.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (ZenGround0) [reply chain/stmgr/forks.go:1250]: I defer to you and @arajasek

File: extern/storage-sealing/precommit_batch.go [hunk 21]
```diff
@@ -0,0 +1,312 @@
...
+			sendAboveMax = true
+		case <-b.batchWait(cfg.PreCommitBatchWait, cfg.PreCommitBatchSlack):
+			sendAboveMin = true
+		case fr := <-b.force: // user triggered
+			forceRes = fr
+		}
+
+		var err error
+		lastMsg, err = b.processBatch(sendAboveMax, sendAboveMin)
+		if err != nil {
```
Comment (GFZRZK) [inline extern/storage-sealing/precommit_batch.go:106]: Simply handle errors here will cause a deadlock. Amusing that we meet error in `processBatch` i.e.: an API error or meet no enough money. And it lasts till sector expired, then here will always return error from `sendMsg` since specs-actors will not allow a expired sector pass. I think when meet error, here should check every sector's validation and close and send the error to those sectors which will never success

Comment (GFZRZK) [reply extern/storage-sealing/precommit_batch.go:106]: @magik6k what's your opinion?
